src/core/qgstemporalnavigationobject.cpp

```diff
@@ -0,0 +1,183 @@
+/***************************************************************************
+                         qgstemporalnavigationobject.cpp
+                         ---------------
+    begin                : March 2020
+    copyright            : (C) 2020 by Samweli Mwakisambwe
+    email                : samweli at kartoza dot com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgstemporalnavigationobject.h"
+#include "qgsproject.h"
+#include "qgsprojecttimesettings.h"
+#include "qgsmaplayer.h"
+#include "qgsrasterlayer.h"
+
+QgsTemporalNavigationObject::QgsTemporalNavigationObject()
+{
+}
+
+QDateTime QgsTemporalNavigationObject::addToDateTime( QDateTime datetime, QString time, int value )
+{
+  if ( time == QString( "Minutes" ) )
+  {
+    return datetime.addSecs( value * 60 );
+  }
+  if ( time == QString( "Hours" ) )
+  {
+    return datetime.addSecs( value * 3600 );
+  }
+  if ( time == QString( "Days" ) )
+  {
+    return datetime.addDays( value );
+  }
+  if ( time == QString( "Months" ) )
+  {
+    return datetime.addMonths( value );
+  }
+  if ( time == QString( "Years" ) )
+  {
+    return datetime.addYears( value );
+  }
+
+  return datetime;
+}
+
+void QgsTemporalNavigationObject::updateLayersTemporalRange( QDateTime datetime, QString time, int value )
```

Comment:
Drop this method -- we should **never** directly set the layer temporal properties from a time controller, and instead we must always go through the canvas -> map settings -> render context -> renderer path.